plugins/inputs/vsphere/vsan.go

```diff
@@ -0,0 +1,523 @@
+package vsphere
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/methods"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+	vsanmethods "github.com/vmware/govmomi/vsan/methods"
+	vsantypes "github.com/vmware/govmomi/vsan/types"
+)
+
+const (
+	vsanNamespace = "vsan"
+	vsanPath      = "/vsanHealth"
+	hwMarksKey    = "vsan-perf"
+	perfPrefix    = "performance."
+)
+
+var (
+	vsanPerfMetricsName    string
+	vsanSummaryMetricsName string
+	perfManagerRef         = types.ManagedObjectReference{
+		Type:  "VsanPerformanceManager",
+		Value: "vsan-performance-manager",
+	}
+	healthSystemRef = types.ManagedObjectReference{
+		Type:  "VsanVcClusterHealthSystem",
+		Value: "vsan-cluster-health-system",
+	}
+	spaceManagerRef = types.ManagedObjectReference{
+		Type:  "VsanSpaceReportSystem",
+		Value: "vsan-cluster-space-report-system",
+	}
+)
+
+// collectVsan is the entry point for vsan metrics collection
+func (e *Endpoint) collectVsan(ctx context.Context, resourceType string, acc telegraf.Accumulator) error {
+	if versionLowerThan(e.apiVersion, "5.5") {
+		log.Printf("I! [inputs.vsphere][vSAN] Minimum API Version 5.5 required for vSAN. Found: %s. Skipping VCenter: %s", e.apiVersion, e.URL.Host)
+		return nil
+	}
+	vsanPerfMetricsName = strings.Join([]string{"vsphere", "vsan", "performance"}, e.Parent.Separator)
+	vsanSummaryMetricsName = strings.Join([]string{"vsphere", "vsan", "summary"}, e.Parent.Separator)
+	res := e.resourceKinds[resourceType]
+	client, err := e.clientFactory.GetClient(ctx)
+	if err != nil {
+		return fmt.Errorf("fail to get client when collect vsan: %v", err)
+	}
+	// Create vSAN client
+	vimClient := client.Client.Client
+	vsanClient := vimClient.NewServiceClient(vsanPath, vsanNamespace)
+	if e.Parent.TLSCA != "" {
+		if err := vsanClient.SetRootCAs(e.Parent.TLSCA); err != nil {
+			return err
+		}
+	}
+	// vSAN Metrics to collect
+	metrics := e.getVsanMetadata(ctx, vsanClient, res)
+	// Iterate over all clusters, run a goroutine for each cluster
+	te := NewThrottledExecutor(e.Parent.CollectConcurrency)
+	for _, obj := range res.objects {
+		te.Run(ctx, func() {
+			e.collectVsanPerCluster(ctx, obj, vimClient, vsanClient, metrics, acc)
+		})
+	}
+	te.Wait()
+	return nil
+}
+
+// collectVsanPerCluster is called by goroutines in collectVsan function.
+func (e *Endpoint) collectVsanPerCluster(ctx context.Context, clusterRef objectRef, vimClient *vim25.Client, vsanClient *soap.Client, metrics map[string]string, acc telegraf.Accumulator) {
+	// Construct a map for cmmds
+	cluster := object.NewClusterComputeResource(vimClient, clusterRef.ref)
+	cmmds, err := getCmmdsMap(ctx, vimClient, cluster)
```

Comment:
If you need `cmmds` only on line 107, move this function down near there so that what you are doing is localized.  This will make it easier to follow the logic of why this is called.